src/auth/src/credentials/internal/sts_exchange.rs

```diff
@@ -0,0 +1,442 @@
+}
+
+#[allow(dead_code)]
+impl STSHandler {
+    pub fn new() -> Self {
+        let client = reqwest::Client::new();
+        Self { client }
+    }
+
+    /// performs the token exchange using a refresh token flow with
+    /// the provided [RefreshAccessTokenRequest] information.
+    pub async fn refresh_access_token(
+        &self,
+        req: RefreshAccessTokenRequest,
+    ) -> Result<TokenResponse> {
+        let mut params: HashMap<&str, String> = HashMap::new();
+        params.insert("grant_type", REFRESH_TOKEN_GRANT_TYPE.to_string());
+        params.insert("refresh_token", req.refresh_token);
+
+        self.execute(req.url, req.authentication, req.headers, params)
+            .await
+    }
+
+    /// performs an oauth2 token exchange with the provided [ExchangeTokenRequest] information.
+    pub async fn exchange_token(&self, req: ExchangeTokenRequest) -> Result<TokenResponse> {
+        let mut params: HashMap<&str, String> = HashMap::new();
+
+        params.insert("grant_type", TOKEN_EXCHANGE_GRANT_TYPE.to_string());
+        params.insert("requested_token_type", ACCESS_TOKEN_TYPE.to_string());
+
+        params.insert("subject_token", req.subject_token);
+        params.insert("subject_token_type", req.subject_token_type);
+
+        if !req.scope.is_empty() {
+            params.insert("scope", req.scope.join(" "));
+        }
+
+        if let Some(audience) = req.audience {
+            params.insert("audience", audience);
+        }
+        if let Some(resource) = req.resource {
+            params.insert("resource", resource);
+        }
+        if let Some(actor_token) = req.actor_token {
+            params.insert("actor_token", actor_token);
+        }
+        if let Some(actor_token_type) = req.actor_token_type {
+            params.insert("actor_token_type", actor_token_type);
+        }
+
+        if let Some(options) = req.extra_options {
+            if let Ok(value) = serde_json::to_value(options) {
+                params.insert("options", value.to_string());
+            }
+        }
+
+        self.execute(req.url, req.authentication, req.headers, params)
+            .await
+    }
+
+    /// execute http request and token exchange
+    pub async fn execute(
+        &self,
+        url: String,
+        client_auth: ClientAuthentication,
+        headers: http::HeaderMap,
+        params: HashMap<&str, String>,
+    ) -> Result<TokenResponse> {
+        let mut headers = headers.clone();
+        let mut params = params.clone();
+        println!("[execute] url: {}", url);
+        client_auth.inject_auth(&mut headers, &mut params);
+        println!("[execute] headers: {:?}", headers);
+
+        let res = self
+            .client
+            .post(url)
+            .form(&params)
+            .headers(headers)
+            .send()
+            .await
+            .map_err(|err| {
+                CredentialsError::from_str(false, format!("failed to request token: {}", err))
```

Comment:
You may want to consider enhancing the error handling here. We return OAuthExceptions in Java with a lot more info. We encapsulate more specific error details (like error_code, error_description).

Reply:
I'll create a sub issue to improve that later